### PR TITLE
Sprint4 GitHub version management software

### PR DESCRIPTION
**Pull request by rladdusaw**

### CODE REVIEW

File: src/main/java/edu/tamu/app/model/ServiceType.java [hunk 1]
```diff
@@ -38,6 +39,7 @@ public void setGloss(String gloss) {
     public List<Setting> getScaffold() {
         List<Setting> scaffold = new ArrayList<Setting>();
         switch (this) {
+        case GITHUB:
```
Comment (ghost) [inline src/main/java/edu/tamu/app/model/ServiceType.java:42]: Won’t the front end need the scaffold settings to create a GitHub remote project manager?

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+    public GitHubService(final ManagementService managementService) throws IOException {
+        this.managementService = managementService;
+        github = getGitHubInstance();
+        members = new HashMap<String, Member>();
+    }
+
+    @Override
+    public List<RemoteProject> getRemoteProjects() throws Exception {
+        logger.info("Fecthing remote projects");
```
Comment (ghost) [inline src/main/java/edu/tamu/app/service/manager/GitHubService.java]: Typo...Fetching

Comment (ghost) [reply src/main/java/edu/tamu/app/service/manager/GitHubService.java]: A few more occurrences of this typo below.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        final List<RemoteProject> remoteProjects = new ArrayList<RemoteProject>();
+        final GHOrganization org = github.getOrganization(ORGANIZATION);
+        for (final Entry<String, GHRepository> entry : org.getRepositories().entrySet()) {
+            final List<GHProject> projects = entry.getValue().listProjects().asList();
+            final List<GHLabel> labels = entry.getValue().listLabels().asList();
+            for (final GHProject project : projects) {
```
Comment (ghost) [inline src/main/java/edu/tamu/app/service/manager/GitHubService.java]: A project in Project Management Service should match to GitHub repos, not GitHub projects. GitHub project will map to a sprint as it correctly does below.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        github = getGitHubInstance();
+        members = new HashMap<String, Member>();
+    }
+
+    @Override
+    public List<RemoteProject> getRemoteProjects() throws Exception {
+        logger.info("Fecthing remote projects");
+        final List<RemoteProject> remoteProjects = new ArrayList<RemoteProject>();
+        final GHOrganization org = github.getOrganization(ORGANIZATION);
+        for (final Entry<String, GHRepository> entry : org.getRepositories().entrySet()) {
```
Comment (ghost) [inline src/main/java/edu/tamu/app/service/manager/GitHubService.java]: Can use `values()` of the map and iterate of the `GHRepository`.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+            final List<GHProject> projects = entry.getValue().listProjects().asList();
+            final List<GHLabel> labels = entry.getValue().listLabels().asList();
```
Comment (ghost) [inline src/main/java/edu/tamu/app/service/manager/GitHubService.java]: This is where we can ensure the repo has all the required labels to be a participant in Project Management Service. There may be a better place, but I feel the labels should be validated here to ensure the project will have consistent behavior.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            List<GHProjectCard> projectCards = column.listCards().asList();
+            AtomicInteger cardNumber = new AtomicInteger();
+            for (GHProjectCard card : projectCards) {
+                GHIssue content = card.getContent();
+
+                String id = String.valueOf(card.getId());
+                String name = content.getTitle();
+                String number = String.valueOf(cardNumber.getAndIncrement());
+                // TODO: Figure out what the type priority is
+                String type = "";
```
Comment (ghost) [inline src/main/java/edu/tamu/app/service/manager/GitHubService.java]: `type` should be the value that is to be mapped to internal CardType. The mappings are managed and afford mapping multiple values to a CardType. The type would basically be the label designated for the type; constants defined above.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                String description = content.getBody();
+                String status = card.getColumn().getName();
+                // TODO: Figure out how we want to handle sizes
+                String estimate = null;
```
Comment (ghost) [inline src/main/java/edu/tamu/app/service/manager/GitHubService.java:207]: I would like to propose estimates be labels in which their value can be parsed to a float. `Float.parseFloat(String value)`

Comment (ghost) [reply src/main/java/edu/tamu/app/service/manager/GitHubService.java:207]: Maybe validate them by a set of point estimates we use.

Comment (ghost) [reply src/main/java/edu/tamu/app/service/manager/GitHubService.java:207]: Yeah, simple check if the label value is contained with a static list of points.

**Review by ghost — CHANGES_REQUESTED** [review]
Looks good for the most part. We should keep this project a good example of test coverage. Therefor, needs tests.

**Comment by ghost** [comment]
We will also have to ensure the project/sprint has the columns mapped to the managed status in Project Management Service. Multiple columns can be mapped to a single column, but project/sprint should have the minimal to behave similar to Project Management Service.

File: src/main/java/edu/tamu/app/model/ServiceType.java (hunk 1 above)
Comment (rladdusaw) [reply src/main/java/edu/tamu/app/model/ServiceType.java:42]: This is what might need to change if we have to run it as a GitHub App. For the moment the login details are close enough that I'm just falling through to the same properties as VersionOne. I suppose we don't need the url in this case.

Comment (ghost) [reply src/main/java/edu/tamu/app/model/ServiceType.java:42]: I see.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java (hunk 5 above)
Comment (rladdusaw) [reply src/main/java/edu/tamu/app/service/manager/GitHubService.java]: I think we need to have at least a discussion with the stakeholders before we start modifying every repo we interact with. There was nothing in the planning about doing something like this. It feels like it should be a separate card.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java (hunk 6 above)
Comment (rladdusaw) [reply src/main/java/edu/tamu/app/service/manager/GitHubService.java]: Type is a string on a Card. We need to decide what to do when we get a card from GitHub with multiple valid type labels.

File: src/main/java/edu/tamu/app/service/manager/GitHubService.java (hunk 7 above)
Comment (rladdusaw) [reply src/main/java/edu/tamu/app/service/manager/GitHubService.java:207]: I've sent a write-up to @jcreel and @jmicah detailing the 3 areas where we need to establish a convention. I might schedule a meeting to hash this out next week.

**Review by ghost — APPROVED** [review]

**Pull request merged**